corehq/apps/es/README.rst

```diff
@@ -80,6 +88,113 @@ This reindex procedure is inherently safe because:
   hoster are not outside the Dimagi main track.
 
 
+Design Details
+''''''''''''''
+
+Reindex Procedure Details
+'''''''''''''''''''''''''
+
+1. Configure multiplexing on an index.
+
+   - Configure the document adapter for the index with a "secondary index name".
+     This will cause the adapter to use multiplexing logic instead of a single
+     index.
+
+     **Note**: The multiplexing logic required for this operation is not yet
+     implemented. The multiplexing adapter will most likely delegate to two
+     document adapters configured for separate indexes. Suffice it to say that
+     when a secondary index is defined for an adapter, it effectively becomes a
+     multiplexing adapter (which to the consumer, is indistinguishable from a
+     "standard" adapter).
+
+   - *(Optional)* If the reindex involves other meta-index changes (shards,
+     mappings, etc), also update those configurations at this time.
+   - Add a migration which performs all cluster-level operations required for
+     the new (secondary) index. For example:
+
+     - creates the new index
+     - configures shards, replicas, etc for the index
+     - sets the index mapping
+
+   - Review, merge and deploy this change.  At Django startup, the new
```

Comment:
> The new reindexing workflow will be as safe and painless as any database migration. Reindexing might be more "complex" when it requires the extra "sync" management command, but there shouldn't be any reason to roll out a reindex slowly

For a more concrete example of my concern, what happens when we or, even more pressingly, a third party, needs to reindex because they want to add shards to an existing index? Each individual environment will need to reshard at various times, and it is likely that at some point that may be necessary for every commcare environment but not at the same time. If that operation requires code changes and added migrations, I fear that is impossible for third part hosters to do on their own (as it will both require them to fork the code base and generate migration conflicts for their reindex migrations), and difficult or painful for them (and maybe us) when we need to do that for a given environment.

Reply:
Third parties can't do that now, so I'm considering it "outside the scope of adapters". This design doesn't prohibit that feature, and I'm leaving it to be implemented at a later time.

Reply:
> Third parties can't do that now

I don't think this is quite true. At least for the form and case search indexes the index names can be set via settings which means that an env could create a new index with a new name, use the existing reindex tools (`ptop_reindexer_v2` or `reindex_es_native`) to perform the reindex (or do it manually via the ES API) and then update the settings to point to the new index (and also switch the alias over).

We have certainly done that in the past on individual envs. I think the issue @calellowitz is addressing is that we are tying infrastructure concerns to the code and making it impossible to make transparent infra changes without also making code changes.

For changes that also require code changes e.g. changes to mappings, I think this approach is fine provided it is sufficiently flexible to allow individual envs to determine their own roll-out schedule (within some reasonable time frame).

Reply:
To be clear, this 'hypothetical' scenario is exactly the scenario we were in with ICDS. The 'rec' project also has significant data and are in the process of migrating to their own infra. There are also discussions in USH about deploying CommCare to independent hardware for large EDSS projects (which hopefully won't happen but is not completely unlikely).

I don't think it should be very difficult to make it possible to override these things via settings. This would give us the flexibility to reindex envs independently if necessary.

It could be worth taking this to a doc to flesh out the exact reindex process step by step.